File: q3217/q3217.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode curr = head;

        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) numSet.add(num);

        while (curr != null) {
            if (numSet.contains(curr.val)) {
                prev.next = curr.next; 
                curr = curr.next;
            } else {
                prev = curr; 
                curr = curr.next;
            }
        }

        return dummy.next;
    }
}

File: q3217/q3217.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        numSet = set(nums)
        # Use this dummy to track the head of the link list
        dummy = ListNode(0, head)

        # Set prev to point at dummy, and curr to point at head
        prev = dummy
        curr = head

        while (curr != None):
            # No need to remove node, update prev and curr by moving them forward by 1
            if (curr.val not in numSet):
                prev = curr
                curr = curr.next
            # Need to remove node, prev.next now points to curr.next, then move curr forward by 1
            else:
                prev.next = curr.next
                curr = curr.next
        
        # Original head might not be the correct new head as it might have been removed
        # dummy.next will be pointing to the correct head of the new linked list
        return dummy.next
